File: EaseMall-biz/src/main/java/com/ysmull/easemall/util/UserContext.java
package com.ysmull.easemall.util;

import com.ysmull.easemall.model.entity.User;

/**
 * @author maoyusu
 */
public class UserContext implements AutoCloseable {

    /**
     * 这里故意替换了ThreadLocal，做了个实验
     */
    private static final MyThreadLocal<User> CURRENT = new MyThreadLocal<>();

    public UserContext(User user) {
        if (user != null) {
            CURRENT.set(user);
        }
    }

    public static User getCurrentUser() {
        return CURRENT.get();
    }

    @Override
    public void close() {
        CURRENT.remove();
    }
}

File: EaseMall-biz/src/main/java/com/ysmull/easemall/util/UserContext.kt
package com.ysmull.easemall.util

import com.ysmull.easemall.model.entity.User

/**
 * @author maoyusu
 */
class UserContext(user: User?) : AutoCloseable {

    init {
        if (user != null) {
            CURRENT.set(user)
        }
    }

    override fun close() {
        CURRENT.remove()
    }

    companion object {

        /**
         * 这里故意替换了ThreadLocal，做了个实验
         */
        private val CURRENT = ThreadLocal<User>()

        val currentUser: User
            get() = CURRENT.get()
    }
}
